docs/architecture/adr/0015-separate-library-per-product.md

```diff
@@ -0,0 +1,72 @@
+---
+parent: Decisions
+nav_order: 15
+adr: "0015"
+status: In progress
+date: 2022-11-03
+tags: [clients]
+---
+
+# Separate library per product
+
+## Context and Problem Statement
+
+Our `clients` repository currently stores much common logic inside of the `libs/common` package,
+which is a library that is consumed by all of our clients. However not all parts of the package are
+actually used by all of our clients which makes our bundles unecessarily large. Another issue with
+this arrangement is the increased cognitive load for the different pods that have to filter out the
+parts of the package that don't concern their work.
+
+## Considered Options
+
+- **Continue as is** - Place all services, models, views, etc in `libs/common`.
+- **Put services that are only used in 1 client inside that client** - This has the benefit of not
+  bloating the common package but will start "polluting" our frontends with request/response models
+  that we're trying to hide behind services. Cognitive load is also not particularly improved.
+  Services that are used in more than 1 client would automatically end up in `common`.
+- **Create new libs for each product** - This involves splitting our common code into `common`,
```

Comment:
the goal should definitely be to mimic team structure, which is a lot of the purpose of the reorg we're going through. I don't like the name product or really even to limit it to some technical boundary. feature is closes, but this is what the `Domain` term was invented for.

We should split into logical feature domains and assign those domains to teams, when a domain is large enough or a team has too many domains, some process of mitosis occurs.

Reply:
If I am interpreting this correctly then I agree! Terminology is hard and my experience is that companies/communities usually have their own dialect, I just haven't learned the one used at Bitwarden yet :)

The part where I seem to have been going in the wrong direction is advocating for using [business capability](https://microservices.io/patterns/decomposition/decompose-by-business-capability.html) as a guideline for identifying packages. Mostly because I looked at the existing teams and this is what I thought was the closest match. With the new team structure that has been flipped entirely on it's head.

I think organizing on feature-domain (or [sub-domain](https://microservices.io/patterns/decomposition/decompose-by-subdomain.html)) is just as valid and we should pick that which matches how we structure our team :)

Reply:
Maybe we should just simplify this by just saying "package per team". But since we seem to have identified many more team-areas than the number of teams we're going to be creating, that might be a mistake?